### PR TITLE
WIP: vault implemention

### PR DESCRIPTION
**Pull request by solokyo**
**WORKING IN PROGRESS, DO NOT MERGE OR USE IN PRODUCTION**
## About this coomit
1. An HTTP server that supports http and https. However, bind port 443 require `sudo setcap CAP_NET_BIND_SERVICE=+eip path/to/pgagroal`. `vault.c` currently is an executable for debugging convenience, and will convert to non-executable if needed.

2. `generate_password` function moved to `security.c` and enhanced.

3. New message added for `vault` - `pgagroal` communication.

4. `rotate_password_cb` added for changing passwords over time.

5. `accpet_vault_cb` and `handle_vault_cb` added for async io between `vault` and `pgagroal`

6. Lots of places are hard coded cause the config is not implemented yet.

## Bug I dealing with:
I want to implement an async listener and handler in `main.c`, there should be a long-lived connection between `pgagroal `and `vault`, and `handle_vault_cb` should read-write messages whenever it received message from `vault`.

But this is what I got from gdb:
```console
Starting program: /home/pgmoneta/pgagroal/build/src/pgagroal -c /etc/pgagroal/pgagroal.conf -a /etc/pgagroal/pgagroal_hba.conf -u /etc/pgagroal/pgagroal_users.conf -F /etc/pgagroal/pgagroal_frontend_users.conf -v

[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[Detaching after fork from child process 32673]
creating thread
/home/pgmoneta/pgagroal/src/libpgagroal/memory.c:111: pgagroal_memory_message: Assertion `message != NULL' failed.

Program received signal SIGABRT, Aborted.
0x00007ffff782abec in __pthread_kill_implementation () from /lib64/libc.so.6
(gdb) bt
#0  0x00007ffff782abec in __pthread_kill_implementation () from /lib64/libc.so.6
#1  0x00007ffff77da956 in raise () from /lib64/libc.so.6
#2  0x00007ffff77c47f4 in abort () from /lib64/libc.so.6
#3  0x00007ffff77c471b in __assert_fail_base.cold () from /lib64/libc.so.6
#4  0x00007ffff77d3536 in __assert_fail () from /lib64/libc.so.6
#5  0x00007ffff79b9ce8 in pgagroal_memory_message () at /home/pgmoneta/pgagroal/src/libpgagroal/memory.c:111
#6  0x00007ffff79bd6ab in read_message (socket=13, block=false, timeout=0, msg=0x7fffffffd558)
    at /home/pgmoneta/pgagroal/src/libpgagroal/message.c:1291
#7  0x00007ffff79b9f53 in pgagroal_read_socket_message (socket=13, msg=0x7fffffffd558)
    at /home/pgmoneta/pgagroal/src/libpgagroal/message.c:87
#8  0x000055555555d679 in handle_vault_cb (loop=0x7ffff7fba060, watcher=0x5555555ad3a0, revents=1)
    at /home/pgmoneta/pgagroal/src/main.c:1978
#9  0x00007ffff7faed0b in ev_invoke_pending () from /lib64/libev.so.4
#10 0x00007ffff7fb28e8 in ev_run () from /lib64/libev.so.4
#11 0x00005555555578f9 in ev_loop (loop=0x7ffff7fba060, flags=0) at /usr/include/ev.h:841
#12 0x000055555555b4c1 in main (argc=10, argv=0x7fffffffdcd8) at /home/pgmoneta/pgagroal/src/main.c:1210
(gdb) 
```

I don't know why `handle_vault_cb` is called, I didn't send anything from `vault` to `pgagroal`. 

Also, I am unsure if I am doing the message read-write right.

### CODE REVIEW

**Comment by jesperpedersen** [comment]
Let it go to a separate process.

We just need to update the frontend passwords and then serve them to the vault binary

File: src/vault.c [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+int main(int arvc, char** argv) {
+    initialize_ssl();
+    ssl_ctx = create_ssl_context();
+    configure_ssl_context(ssl_ctx);
+
+    printf("creating thread\n");
+
+    client_socket = socket(AF_INET, SOCK_STREAM, 0);
+    if (client_socket == -1) {
+        perror("Socket creation error");
```
Comment (fluca1978) [inline src/vault.c:301]: errx

File: src/vault.c [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        exit(EXIT_FAILURE);
+    }
+    
+    struct sockaddr_in serverAddress;
+    serverAddress.sin_family = AF_INET;
+    serverAddress.sin_port = htons(6789);
+    serverAddress.sin_addr.s_addr = inet_addr("127.0.0.1");
+    
+    if (connect(client_socket, (struct sockaddr *)&serverAddress, sizeof(serverAddress)) == -1) {
+        perror("Connection error");
```
Comment (fluca1978) [inline src/vault.c:311]: errx

File: src/vault.c [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        close(client_socket);
+        exit(EXIT_FAILURE);
+    }
+
+    pthread_t http_thread, https_thread;
+
+    int result = pthread_create(&http_thread, NULL, http_thread_func, NULL);
+    if(result != 0) {
+        printf("something wrong\n");
```
Comment (fluca1978) [inline src/vault.c:320]: warnx

**Comment by fluca1978** [comment]
I do agree in not using thread right here.
Also, as a minor note, you should run crustify against the code.

File: src/vault.c (hunk 3 above)
Comment (solokyo) [reply src/vault.c:320]: Thank you for your feedback. I am aware that there are many flaws in the code I submitted in the PR. I opened this PR to discuss my implementation with you all. Once we have finalized how to implement these features, I will close this PR and submit a clean one.

**Comment by solokyo** [comment]
> Let it go to a separate process.

What do you mean by "it"?

> We just need to update the frontend passwords and then serve them to the vault binary

Are you suggesting `pgagroal` keep sending updated passwords to the `vault` and `vault` will keep a cache of it? When client comes and ask for password, `vault` don't need to talk to `pgagroal`.

**Comment by jesperpedersen** [comment]
No, let `pgagroal` update the front-end passwords. And then, the `vault` process ask for the current password whenever there is a client request for it

File: src/vault.c [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+    }
+    pthread_create(&https_thread, NULL, https_thread_func, NULL);
```
Comment (jesperpedersen) [inline src/vault.c:325]: We should follow the process model instead of using threads

File: src/main.c [hunk 5]
```diff
@@ -1043,6 +1055,54 @@ main(int argc, char** argv)
...
+       * 1. start rotate cb
+       * 2. start async vault accept cb
+       * 3. fork exec vault
+       */
+      pid_t pid;
+
+      ev_periodic_init(&rotate_frontend_password, rotate_frontend_password_cb, 0., 60, 0);
+      ev_periodic_start (main_loop, &rotate_frontend_password);
+
+      unix_vault_socket = socket(AF_INET, SOCK_STREAM, 0);
```
Comment (jesperpedersen) [inline src/main.c:1069]: Why do we need this ? The request for the current frontend password can go over the management port...

**Comment by ashu3103** [comment]
@jesperpedersen, I have gone through the task and the corresponding files and I wish to continue this task, kindly assign me.

### Things to be fixed as per discussions

#### pgagroal

- Inside **pgagroal**, the valut's requests will be directed to management port.
- Incorporate some way in **pgagroal** `main.c` to start **pgagroal-vault** server (by running `pgagroal-vault` executable in a seperate child process).
- The logic to trigger `rotate_frontend_password_cb` is run periodically after X seconds (uisng ev_periodic watcher) where X is to be taken from configuration.

#### pgagroal-vault

- The host can be same as the host of pgagroal and can be accessed from configuration and the port is defined 80 for HTTP and 443 for HTTPS.

### Doubts

- Suppose user_a and user_b are two frontend users and user_a sends an HTTP request `http://XXX:YYY/user_b` to the vault so will it get the password of user_b? I mean how are we gonna authenticate the users in case of HTTP.
- Still not sure where to implement the command to change the frontend_user passwords, in vault or in pgagroal?

### Any other suggestions...

**Comment by jesperpedersen** [comment]
In pgagroal there should be a `GET_PASSWORD` management command. pgagroal should not "talk" to pgagroal-vault in any way. pgagroal should handle the changing of the frontend end user passwords.

pgagroal-vault is its own binary, and is independent of pgagroal.

The port configuration should be defined in `pgagroal-vault.conf`, and for the initial implementation there is no need for HTTPS support nor user authentication.

Think of pgagroal-vault as a "proxy" that speaks HTTP to get the current front end password for a specific user.

**Comment by ashu3103** [comment]
> In pgagroal there should be a `GET_PASSWORD` management command. pgagroal should not "talk" to pgagroal-vault in any way. pgagroal should handle the changing of the frontend end user passwords.
> 

Right, so pgagroal-vault should connect to management port of pgagroal which then handles the `GET_PASSWORD` request using messages exchanges.

> pgagroal-vault is its own binary, and is independent of pgagroal.
> 

Yeah, but it has to be spawned by pgagroal in the initialization stage of pgagroal if `enable_vault` (configuration parameter) is true.

> The port configuration should be defined in `pgagroal-vault.conf`

Can we have a more specific format of `pgagroal-vault.conf`.

>and for the initial implementation there is no need for HTTPS support nor user authentication.
> Think of pgagroal-vault as a "proxy" that speaks HTTP to get the current front end password for a specific user.

Got it!

**Comment by ashu3103** [comment]
> `pgagroal-vault.conf` will just have the configuration parameters needed for that process.

I am thinking `pgagroal-vault.conf`, should be like this...

```
host
management-port  # of pgagroal
max_connections
tls
tls_cert_file
tls_key_file
tls_ca_file

```

**Comment by jesperpedersen** [comment]
Yeah, something like

```
[pgagroal-vault]
host = localhost
port = 2500

[main]
host = localhost
port = 2345
user = admin
#tls
#tls_cert_file
#tls_key_file
#tls_ca_file
```

For now, 1 server connection definition is enough...

**Comment by jesperpedersen** [comment]
Work is being done on #407 now

**Pull request closed without merging**